README.md

Use AWS Lambda to expire EBS snapshot with flexible retention rules.

## Why Use Lambda to expire snapshots?

Some EBS snapshot management tools combine creating and expiring snapshots into a single tool.

Snapshot creation should be done from the host, where you have the flexibility to do things like freeze the filesystem or pause a service during the freeze.

However, deleting snapshots should *not* be done from hosts. When that pattern is used, it means a comprimise of a server can lead to deletion of the backups of the server as well.

AWS Lambda is very inexpensive and runs independently of any host

## Why use this snapshot deletion tool?

  * Flexible retention rules
  * Test coverage
  * Multiple retention policies using a single Lambda function.
  * Multiple region support

Our flexible retention rules are implemented by the [grandfatherson](https://github.com/RideAmigosCorp/grandfatherson) module, which has it's own test coverage to confirm the algorithm is working as expected.

The code here is also split into functions that are unit tested, expecting where we actually make the AWS calls.

Deleting the right data and not too much of it is incredibly important feature for a tool that deletes backups. Yet many tools that offer to help you delete your snapshots have no test coverage to confirm that they are working as expected.

## Usage

Use this command to create a .zip file to upload to AWS Lambda.

   npm run package-for-deploy

Create a new Lambda function in the AWS management console or via CLI and choose "Upload a .zip file" for the code entry type.

Make sure an IAM role with the apporopriate snapshot policy is available for the execution role.

To create a trigger, visit AWS Cloudwatch and create a new event rule

Select the "cron" option and express the schedule you want to run your snapshots on. For example, every day at 3 AM UTC would look like this:


    0 3 * * ? *

Choose to customize the event to send "Constant JSON". This is where your express
your configuration. An example is in [./example-event.json](./example-event.json)
which looks like this:

```json
    {
      "dryRun" : true,
      "retentionRules" : {
         "days"   : 14,
         "weeks"  : 13,
         "months" : 12
      },
      "filters": [ { "Name" : "tag:Name", "Values" : ["automated-backup" ] } ],
      "regions" : ["us-east-1","ca-central-1"],
    }
```

As would you guess, `dryRun` produces logging output, but takes no action. Highly recommended to try this first. For all the options for `retentionRules`,
see [grandfatherson](https://github.com/RideAmigosCorp/grandfatherson).

The `filters` are used to select which snapshots you want to prune. For the full list of options see the [docs for the AWS EC2 DescribeVolumes](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVolumes.html) API call.

`regions` is used to set the regions you want expire your snapshots in. One Lambda function can expire snapshots in other regions besides where the Lambda is hosted. The default value is the region where the Lambda is running.

In the above example, we are searching for all snapshots where the tag "Name" has the value 'automated-backup'

### Implementing multiple retention rules

Say you want to apply a different retention policy to some backups, like your database data. There are two simple steps to do that:

  1. Use a different `filters` to find them. For example, give them a different tag that you can search for.
  2. Use CloudWatch to create a second Event Rule that's attached to the same AWS Lambda function.

## Testing / Dry Run

After doing a dry-run, you can view the related log stream in CloudWatch to confirm that you would be retaining and deleting what you expected. The diagnotics could improved here. Patches welcome.

## Logging

During live runs, we log the IDs of snapshots deleted to CloudWatch. Patches would be welcome to create more detailed logging that might include the related volume-ID or other details of the snapshots deleted.

## Alternatives

After this was developed, [AWS Backup](https://aws.amazon.com/backup/) was
released and provides similar functionality. AWS Backup assigns lifecycle rules
to snapshots as they are created. The "AWS Backup" approach does not require a
rotation Lamba like this. The downside to the AWS Backup approach is that if
you change your expiration rules, it won't retro-actively apply them to backups
that have already been created. That can be a bummer if you accidentally create too many backups.

Using this tool, when you update the retention rules existing snapshots will be affected.

## Contributing

Bug fixes and feature contributions are welcome, just be prepared to implement your own feature requests.

## Author

Mark Stosberg <mark@rideamigos.com>

## LICENSE

Apache 2. See [LICENSE](./LICENSE)




CHANGELOG.md

# Change Log

## [1.2.1] - 2019-08-15

### Internals 

 * Improve README, mention AWS Backup

### Packaging

 * Update dependencies

## [1.2.0] - 2017-10-08

### Bug Fixes

 * Fix bug with multi-region support that could cause
   "InvalidSnapshot.NotFound: The snapshot 'snap-...' does not exist." (@pserwylo)

## [1.1.2] - 2017-07-18

### Packaging

 * Fixed dependency warning with newer NPM (@fidian)
 * Bumped dependency versions (@fidian)

### Internals

 * No longer emit "ERROR" to logs when there are no errors (@fidian)

## [1.1.1] - 2017-07-10

### Docs

 * Fixed JSON syntax in example (@fidian)

## [1.1.0] - 2016-12-20

### Added

 * Multiple region support (@markstos)

## [1.0.0]

Initial release.


test/index.js

var index = require('../index');
var _ = require('lodash');
var chai = require('chai');
var moment = require('moment');
var expect = chai.expect;

describe('organizeSnapshotsByVolume', function() { var organizeSnapshotsByVolume = index.organizeSnapshotsByVolume;

  it("single snapshot should be transformed as expected", function () {
    // This is taken straight from the AWS docs
    var snapshots = [
        {
            "Description": "This is my snapshot.",
            "VolumeId": "vol-049df61146c4d7901",
            "State": "completed",
            "VolumeSize": 8,
            "Progress": "100%",
            "StartTime": "2014-02-28T21:28:32.000Z",
            "SnapshotId": "snap-1234567890abcdef0",
            "OwnerId": "012345678910"
        }
    ];

    var byVolume = organizeSnapshotsByVolume(snapshots);
    expect(byVolume).to.eql({
        "vol-049df61146c4d7901" : [
            { SnapshotId: "snap-1234567890abcdef0", StartTime: "2014-02-28T21:28:32.000Z" }
        ]
    });
  });

  it("two snapshots for different volumes should be transformed as expected", function () {
    // This is taken straight from the AWS docs
    var snapshots = [
        {
            "Description": "This is my snapshot.",
            "VolumeId": "vol-049df61146c4d7901",
            "State": "completed",
            "VolumeSize": 8,
            "Progress": "100%",
            "StartTime": "2014-02-28T21:28:32.000Z",
            "SnapshotId": "snap-1234567890abcdef0",
            "OwnerId": "012345678910"
        },
        {
            "Description": "This is my snapshot.",
            "VolumeId": "vol-049df61146c4d7902",
            "State": "completed",
            "VolumeSize": 8,
            "Progress": "100%",
            "StartTime": "2014-02-28T21:28:32.001Z",
            "SnapshotId": "snap-1234567890abcdef1",
            "OwnerId": "012345678910"
        }
    ];

    var byVolume = organizeSnapshotsByVolume(snapshots);
    expect(byVolume).to.eql({
        "vol-049df61146c4d7901" : [ { SnapshotId: "snap-1234567890abcdef0", StartTime: "2014-02-28T21:28:32.000Z" } ],
        "vol-049df61146c4d7902" : [ { SnapshotId: "snap-1234567890abcdef1", StartTime: "2014-02-28T21:28:32.001Z" } ]
    });

  });
});

describe('selectSnapsToDelete', function() {
  var selectSnapsToDelete = index.selectSnapsToDelete;
  it("should select oldest day for deletion when there are snaps for last 3 days and retention rule says to keep one day", function () {

    var snapshots = [
        { SnapshotId: "snap-1234567890abcdef0", StartTime: new Date("1999-12-31T21:28:32.000Z") },
        { SnapshotId: "snap-1234567890abcdef1", StartTime: new Date("1999-12-30T21:28:32.000Z") },
        { SnapshotId: "snap-1234567890abcdef2", StartTime: new Date("1999-12-29T21:28:32.000Z") },
    ];

    var retentionRules = {
      now  : moment.utc('1999-12-31'),
      days : 2
    }

    var toDelete = selectSnapsToDelete(snapshots, retentionRules).toDelete;

    expect(toDelete).have.lengthOf(1);
    expect(toDelete).to.eql([
        { SnapshotId: "snap-1234567890abcdef2", StartTime: new Date("1999-12-29T21:28:32.000Z") },
    ])

  });

 it("should pass real world test with lots of snapshots and complex rentention rules", function () {

   var snapshots =[ { "SnapshotId": "snap-e95b2785", "StartTime": new Date("2016-09-08T02:01:03.000Z") } ,
                  { "SnapshotId": "snap-a954f0c7", "StartTime":   new Date("2016-10-15T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-2b836c4f", "StartTime":   new Date("2016-09-17T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-9b0431e2", "StartTime":   new Date("2016-09-27T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-ec4f9394", "StartTime":   new Date("2016-09-30T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-445d6c3c", "StartTime":   new Date("2016-10-03T02:01:03.000Z") } ,
                  { "SnapshotId": "snap-b1706fb3", "StartTime":   new Date("2016-09-18T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-d92c6fdc", "StartTime":   new Date("2016-10-14T02:01:03.000Z") } ,
                  { "SnapshotId": "snap-cd5b23cb", "StartTime":   new Date("2016-09-07T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-22b6493f", "StartTime":   new Date("2016-09-06T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-4b885257", "StartTime":   new Date("2016-10-01T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-d7e848cb", "StartTime":   new Date("2016-09-24T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-f18bdfed", "StartTime":   new Date("2016-10-11T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-5727014b", "StartTime":   new Date("2016-09-26T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-b5d551aa", "StartTime":   new Date("2016-09-25T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-afc0e5b1", "StartTime":   new Date("2016-09-19T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-b4157fa3", "StartTime":   new Date("2016-09-01T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-fb53c4ed", "StartTime":   new Date("2016-09-15T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-13063c05", "StartTime":   new Date("2016-09-20T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-1637aafd", "StartTime":   new Date("2016-09-16T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-99fa7a78", "StartTime":   new Date("2016-10-08T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-2f73d9cd", "StartTime":   new Date("2016-09-10T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-f332ff17", "StartTime":   new Date("2016-09-02T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-8b95e76f", "StartTime":   new Date("2016-08-31T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-11fc57f6", "StartTime":   new Date("2016-09-11T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-07ae8de0", "StartTime":   new Date("2016-10-06T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-e1537950", "StartTime":   new Date("2016-10-10T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-8c29393f", "StartTime":   new Date("2016-10-20T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-c118ed73", "StartTime":   new Date("2016-10-07T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-1a519da8", "StartTime":   new Date("2016-10-05T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-fc5f8f4e", "StartTime":   new Date("2016-09-23T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-b323d206", "StartTime":   new Date("2016-09-22T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-74d173c1", "StartTime":   new Date("2016-10-16T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-f9d8aa4d", "StartTime":   new Date("2016-10-17T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-3938a28e", "StartTime":   new Date("2016-10-18T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-214a6496", "StartTime":   new Date("2016-09-29T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-95e79d23", "StartTime":   new Date("2016-09-28T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-ae170718", "StartTime":   new Date("2016-10-19T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-e7fa996e", "StartTime":   new Date("2016-10-02T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-a71c5f2e", "StartTime":   new Date("2016-10-04T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-213aeea9", "StartTime":   new Date("2016-10-21T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-542463df", "StartTime":   new Date("2016-09-21T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-8737120c", "StartTime":   new Date("2016-10-09T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-8160b90d", "StartTime":   new Date("2016-10-12T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-fbba7463", "StartTime":   new Date("2016-09-03T02:01:03.000Z") } ,
                  { "SnapshotId": "snap-f507e56f", "StartTime":   new Date("2016-09-14T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-327438a8", "StartTime":   new Date("2016-09-04T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-1423b489", "StartTime":   new Date("2016-09-13T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-0cf50990", "StartTime":   new Date("2016-09-12T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-384100a4", "StartTime":   new Date("2016-09-09T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-2b429bb4", "StartTime":   new Date("2016-10-13T02:01:02.000Z") } ,
                  { "SnapshotId": "snap-c7853658", "StartTime":   new Date("2016-09-05T02:01:02.000Z") } ];

    var retentionRules = {
     "days"   : 14,
     "weeks"  : 13,
     "months" : 12,
     "years"  : 1
    }

    var toDelete = selectSnapsToDelete(snapshots, retentionRules).toDelete;

    // Just reality check the the number of snapshots we start with.
    expect(snapshots).to.have.lengthOf(52);

   expect(toDelete).to.have.lengthOf(31);
 });
});


describe('delParams', function() {
  var delParams = index.delParams;
  it("should replace StartTime with dryRun passed in", function () {
    var snapshots = [
        { SnapshotId: "snap-1", StartTime: "1999-12-31T21:28:32.000Z" },
        { SnapshotId: "snap-2", StartTime: "1999-12-30T21:28:32.000Z" },
    ];

    var dryRun = true;

    var params = delParams(snapshots, dryRun);

    expect(params).to.eql([
        { SnapshotId: "snap-1", DryRun: true },
        { SnapshotId: "snap-2", DryRun: true },
    ])

  });
});


index.js
"use strict";
var async  = require('async');
var moment = require('moment');
var assert = require('assert');
var aws    = require("aws-sdk");
var gfs    = require('grandfatherson');
var _      = require("lodash");


class ExpireSnapshots {
  // Main logic for the Lamda event handler
  static handler (event, context, handlerCallback) {

    // If you don't provide a regions array, default to the region Lambda is running in.
    if (!event.regions) {
      event.regions = [process.env.AWS_REGION];
    }

    // Using CloudWatch, The config is passed as an event using the "Constant (JSON text)" option
    const config = event;

    // validate event
    if (!config.hasOwnProperty('filters')) {
      return handlerCallback('filters must be passed in event.  ex:  "filters": [ { "Name": "tag:Name", "Values": [ "automated-backup" ] } ]');
    }
    if (!config.hasOwnProperty('dryRun')) {
      return handlerCallback("config.dryRun is required.");
    }

    if (!config.hasOwnProperty('retentionRules')) {
      return handlerCallback("config.retentionRules is required.");
    }

    // Run the snapshots in all configured regions
    config.regions.map(function (region) {
      console.log("START snapshot expiration for region: "+region);

      var ec2 = new aws.EC2({apiVersion: "2016-09-15", "region": region});

      // Use filters to find a list of volume Ids to pass to ec2-expire-snapshots
      // We always want DryRun:false here, because describing the snapshots is a safe, read-only operation
      ec2.describeSnapshots({ DryRun: false, Filters: config.filters }, function(err, data) {
        if (err) {
          return handlerCallback(err, err.stack);
        }

        var byVolume = ExpireSnapshots.organizeSnapshotsByVolume(data.Snapshots);

        var filteredSnaps = [];

        // For each volume ID, delete volumes according to retention rules.
        // If dryRun was provided, don't delete just return what we would do.
        // When we are done, callback to note we are done with the Lambda function
        async.mapSeries(Object.keys(byVolume), function (volId, volumeCallback) {
          var filteredSnaps = ExpireSnapshots.selectSnapsToDelete( byVolume[volId], config.retentionRules );

          // Not Debugging! Leave this in.
          console.info("For %s keeps snaps with these relative ages %j",volId,filteredSnaps.toKeep.reverse());

          // Delete all snapshots, possibly  with the dry run flag,
          // then callback to note we are done with the current volume
          var snapsToDeleteParams  = ExpireSnapshots.delParams(filteredSnaps.toDelete, config.dryRun);
          async.mapSeries(snapsToDeleteParams, ExpireSnapshots.delOneSnap(ec2), volumeCallback);
        }, function (err, results) {
          if (err) {
            console.log("ERROR");
            console.dir(err)
          } else {
            console.log("SUCCESS");
          }

          console.log("RESULTS");
          console.dir(results)
          console.log("END snapshot expiration for region: "+region);
          handlerCallback();
        });
      });
    })
  }

  /*
    Re-organize the snapshots based on volume IDs.
    Given 'Snapshots' array as returned from EC2 API,
    Create an object with just the SnapshotId and StartTime picked
    {
       "vol-049df61146c4d7901" : [ { SnapshotId: "snap-1234567890abcdef0", StartTime: "2014-02-28T21:28:32.000Z"  }, ... ],
      ...
    }
  */
  static organizeSnapshotsByVolume (snapshots) {
    var byVolume = {};

    snapshots.map(function (snap) {
        if (!byVolume[snap.VolumeId]) {
          byVolume[snap.VolumeId] = [];
        }

        byVolume[snap.VolumeId].push(_.pick(snap, ['SnapshotId','StartTime']));
    });

    return byVolume;

  } 

  // Given an array of snapshots
  // given as { SnapshotId: snapId : StartTime: 'createdDate' } and grandfatherson retention rules,
  // filter list to those would be deleted.
  // Object returned contains both 'toDelete' array with filtered dates and `toKeep` array with just date strings
  // The toKeep array contains humanized-strings-- not dates-- for debugging / dry runs to manually confirm correctness

  static selectSnapsToDelete (snapshots, retentionRules) {
    var filteredSnaps = [];

    var datesToDelete = gfs.toDelete( _.map(snapshots,'StartTime'), retentionRules );

    var datesToDeleteAsISO = datesToDelete.map(function (dt) { return dt.toISOString()  } );

     // If the snapshot date matches one of the dates to delete, mark it for deletion
     var snapsToDelete = _.filter(snapshots, function (snapshot) {
        // Remember that indexOf can return zero as match
        if (datesToDeleteAsISO.indexOf(snapshot.StartTime.toISOString()) >= 0) {
          return true
        }
     });

    var datesToKeep = gfs.toKeep( _.map(snapshots,'StartTime'), retentionRules );
    // Return human-friendly ages of snapshots to keep.
    // These dates are relative the newest snapshot.
    var datesToKeepHumanized = datesToKeep.map(function (dt) {
        return moment.duration(moment.utc(_.last(datesToKeep)).diff(dt)).humanize()+' ago'
    });

    return { 
        toDelete: snapsToDelete,
        toKeep: datesToKeepHumanized
    }
  }

  // Given array of objects which contain SnapshotId, transform them into array of params
  // to pass to ec2.deleteSnapshot, factoring config.dryRun setting
  static delParams (snapshots, dryRun) {
    return _.chain(snapshots)
            .map(function(o) { return _.pick(o, "SnapshotId"); })
            .map(function(o) { return _.extend(o, { DryRun: dryRun }); })
            .value();
  }

  // Returns async map iterator, which doesn't depend on the global ec2 object.
  static delOneSnap(ec2) {
    return function(params, callback) {
      const PAUSE_MILLISECONDS = 100; // Avoid AWS throttling.
      setTimeout(function () {
         ec2.deleteSnapshot(params, function(err, data) {
           if (err) {
             if (err.hasOwnProperty("code") && err.code == "DryRunOperation") {
               //data = err;
               data = 'DRY RUN: Would delete snapshot-id '+params.SnapshotId;
               err = null;
             }
             callback(err, data);
           }
           // Deleting was successful and it was not a dry run. Log and callback
           else {
             callback(null, data);
           }
       });
      }, PAUSE_MILLISECONDS);
    }
  }
}

module.exports = ExpireSnapshots;
